semgrep-core/toy-matcher/Matcher.ml

```diff
@@ -0,0 +1,562 @@
+    let orig_pat = pat in
+    let orig_input = input in
+    let in_ellipsis = ellipsis <> None in
+    match pat with
+    | [] ->
+        (match input with
+         | [] ->
+             let env = close_ellipsis ellipsis env in
+             Some env
+         | symbol :: input ->
+             if in_ellipsis then
+               let ellipsis = extend_ellipsis ellipsis symbol in
+               match_ ellipsis env orig_pat input
+             else
+               None
+        )
+    | (pat_atom, opt_name) :: pat ->
+        match input with
+        | [] ->
+            (* end of input, only empty sequence can match *)
+            (match pat_atom with
+             | Ellipsis ->
+                 let env = extend env ellipsis opt_name [] in
+                 match_ (init_ellipsis opt_name []) env pat input
+             | Backref name ->
+                 (match Env.find_opt name env with
+                  | Some [] ->
+                      let env = extend env ellipsis opt_name [] in
+                      match_ None env pat input
+                  | _ ->
+                      None
+                 )
+             | _ -> None
+            )
+        | symbol :: input ->
+            let head_match =
+              match pat_atom with
+              | Any_symbol ->
+                  let env = extend env ellipsis opt_name [symbol] in
+                  Some (None, env, input)
+              | Symbol x ->
+                  if x = symbol then
+                    let env = extend env ellipsis opt_name [symbol] in
+                    Some (None, env, input)
+                  else
+                    None
+              | Ellipsis ->
+                  (* start new ellipsis but don't consume first symbol *)
+                  let env = close_ellipsis ellipsis env in
+                  Some (init_ellipsis opt_name [], env, orig_input)
+              | Backref name ->
+                  (match Env.find_opt name env with
+                   | Some [symbol0] when symbol0 = symbol ->
+                       let env = extend env ellipsis opt_name [symbol] in
+                       Some (None, env, input)
+                   | _ ->
+                       None
+                  )
+            in
+            let matches_here =
+              match head_match with
+              | None -> None
+              | Some (ellipsis, env, input) ->
+                  (* match the rest of the pattern with the rest of the
+                     input *)
+                  match_ ellipsis env pat input
+            in
+            match matches_here with
+            | Some _ as x -> x
+            | None ->
+                if in_ellipsis then
+                  (* skip input symbol and retry *)
+                  let ellipsis = extend_ellipsis ellipsis symbol in
+                  match_ ellipsis env orig_pat input
+                else
+                  None
+  in
+  if cache then
+    get_from_cache := Memoize.create uncached_match
```

Comment:
Would be great to have an example where I can understand how memoizing would improve the performance.
On your sample_pattern, with a simple input like A B C C C C C C C C B C C
why do we avoid redoing the same work thx to memoization?

Reply:
This is a lot harder to explain than I thought. See other my comment. Best illustration is a trace.